std/typecons.d

```diff
@@ -3909,6 +3909,335 @@ if (is (typeof(nullValue) == T))
+        assert(absent != a);
+        assert(absent != b);
+        assert(absent != c);
+
+        assert(a == b);
+        assert(a != c);
+
+        assert(absent != 42);
+        assert(a == 42);
+        assert(c != 42);
+    }
+
+    @safe unittest
+    {
+        // Test constness
+        immutable Optional!int a = 42;
+        Optional!int b = 42;
+        immutable Optional!int c = 29;
+        Optional!int d = 29;
+        immutable e = 42;
+        int f = 29;
+        assert(a == a);
+        assert(a == b);
+        assert(a != c);
+        assert(a != d);
+        assert(a == e);
+        assert(a != f);
+
+        // Test rvalue
+        assert(a == const Optional!int(42));
+        assert(a != Optional!int(29));
+    }
+
+    // Issue 17482
+    @system unittest
+    {
+        import std.variant : Variant;
+
+        Optional!Variant a = Variant(12);
+        assert(a == 12);
+        Optional!Variant e;
+        assert(e != 12);
+    }
+
+    size_t toHash() const @safe nothrow
+    {
+        static if (__traits(compiles, .hashOf(_value.payload)))
+            return _present ? .hashOf(_value.payload) : 0;
+        else
+            // Workaround for when .hashOf is not both @safe and nothrow.
+            return _present ? typeid(T).getHash(&_value.payload) : 0;
+    }
+
+    /**
+     * Gives the string `"Optional.absent"` if `present` is `false`. Otherwise, the
+     * result is equivalent to calling $(REF formattedWrite, std,format) on the
+     * underlying value.
+     *
+     * Params:
+     *     writer = A `char` accepting
+     *     $(REF_ALTTEXT output range, isOutputRange, std, range, primitives)
+     *     fmt = A $(REF FormatSpec, std,format) which is used to represent
+     *     the value if this Optional is present
+     * Returns:
+     *     A `string` if `writer` and `fmt` are not set; `void` otherwise.
+     */
+    string toString()
+    {
+        import std.array : appender;
+        auto app = appender!string();
+        auto spec = singleSpec("%s");
+        toString(app, spec);
+        return app.data;
+    }
+
+    /// ditto
+    string toString() const
+    {
+        import std.array : appender;
+        auto app = appender!string();
+        auto spec = singleSpec("%s");
+        toString(app, spec);
+        return app.data;
+    }
+
+    /// ditto
+    void toString(W)(ref W writer, scope const ref FormatSpec!char fmt)
+    if (isOutputRange!(W, char))
+    {
+        import std.range.primitives : put;
+        if (present)
+            formatValue(writer, _value.payload, fmt);
+        else
+            put(writer, "Optional.absent");
+    }
+
+    /// ditto
+    void toString(W)(ref W writer, scope const ref FormatSpec!char fmt) const
+    if (isOutputRange!(W, char))
+    {
+        import std.range.primitives : put;
+        if (present)
+            formatValue(writer, _value.payload, fmt);
+        else
+            put(writer, "Optional.absent");
+    }
```

Comment:
Why is a non const and const version needed and not just a const version?

Reply:
`toString` should still work if `Optional` embeds a data structure with non-const `toString` (as is, sadly, the default).